### PR TITLE
Course Updating

### PR DESCRIPTION
**Pull request by silverware13**
Add 6784 courses to the database.

Allow the head advisor to update the list of courses in the database to include any new courses from the OSU course catalog.

Add entity that holds a list of all of the required courses (is referenced when the courses in the database are updated).

Change course "credits" column from int to string, as some courses in the course catalog have a range allowed for credits (ex: 1 to 16 credits).

When adding a course with a range of credits, users will be shown a prompt to enter the amount of credits that they would like to pick. An error message will be shown if the user tries to enter an invalid amount of credits.

When loading paginated results the "show more" button no longer needs to be pressed. More results will load as the user scrolls to the bottom of the page.

Handle course filtering on the back-end, using the search courses endpoint.

### CODE REVIEW

File: api/course.js [hunk 1]
```diff
@@ -4,19 +4,63 @@
...
+    // get user data
+    const authenticatedUser = await getUserById(req.auth.userId);
+
+    // due to the computational cost of this action,
+    // only the head advisor can perform this action
+    if (authenticatedUser.role === Role.advisor ||
+        authenticatedUser.role === Role.headAdvisor) {
+
+      // start course update
+      getLiveCourses(req.auth.userId);
```
Comment (philectron) [inline api/course.js:30]: `await`?

File: entities/required.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+const Required = Object.freeze({
+  courses: ["HHS 231", "HHS 241", "COMM 111", "COMM 114", "CS 160", "CS 161",
+    "CS 162", "CS 165", "CS 261", "CS 271", "CS 290", "MTH 231", "CS 225",
+    "MTH 231", "CS 225", "MTH 251", "MTH 252", "WR 121", "WR 214", "WR 222",
+    "ST 314", "WR 327", "CS 325", "CS 340", "CS 344", "CS 352", "CS 361",
+    "CS 362", "CS 372", "CE 372", "ECE 372", "CS 381", "CS 391", "CS 444",
+    "CS 461", "CS 462", "CS 463"]
+});
+exports.Required = Required;
```
Comment (philectron) [inline entities/required.js:12]: Good one. I think this is as close as we can get to Rob's idea of making editing the required courses easier.

File: models/comment.js [hunk 3]
```diff
@@ -14,17 +14,19 @@ async function createComment(planId, userId, text) {
     let results = await pool.query(sql, [planId, userId, text]);
     const commentId = results[0].insertId;
 
+    // get the time that the comment was created
     sql = "SELECT time FROM Comment WHERE commentId=?;";
     results = await pool.query(sql, [commentId]);
+    const time = results[0][0].time;
+
+    // send out notifications about the new comment
+    planNotification(planId, userId, 1);
```
Comment (philectron) [inline models/comment.js:23]: `await`?

File: models/review.js [hunk 4]
```diff
@@ -15,17 +15,19 @@ async function createReview(planId, userId, status) {
     let results = await pool.query(sql, [planId, userId, status, status, planId]);
     const reviewId = results[0][1].insertId;
 
+    // get the time that the review was created
     sql = "SELECT time FROM PlanReview WHERE reviewId=?;";
     results = await pool.query(sql, reviewId);
+    const time = results[0][0].time;
+
+    // send out notifications about the new status
+    planNotification(planId, userId, 2);
```
Comment (philectron) [inline models/review.js:24]: `await`?

File: api/course.js [hunk 5]
```diff
@@ -4,19 +4,63 @@
...
+  try {
+
+    console.log("Get all courses from the OSU course API");
+
+    // get user data
+    const authenticatedUser = await getUserById(req.auth.userId);
+
+    // due to the computational cost of this action,
+    // only the head advisor can perform this action
+    if (authenticatedUser.role === Role.advisor ||
```
Comment (philectron) [inline api/course.js]: Should advisors be able to perform admin operations like updating courses?

File: client/src/components/navbar/Navbar.js [hunk 6]
```diff
@@ -93,15 +90,20 @@ function Navbar() {
...
         ) : (
           null
         )}
-        {role ? <History /> : null}
-        <Notifications />
+        {role === 2 ? (
+          <UpdateCourses />
+        ) : (
+          null
+        )}
```
Comment (philectron) [inline client/src/components/navbar/Navbar.js:106]: image

For some reasons, I couldn't see the `UpdateCourses` button even though I was logged in as a Head Advisor.

The console didn't throw any error.

File: client/src/components/navbar/UpdateCourses.js [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+  }
+
+  return (
+    <div css={style}>
+      <button className="update-courses-button" onClick={() => startCourseUpdate()}>
+        Update Courses
+      </button>
+    </div>
```
Comment (philectron) [inline client/src/components/navbar/UpdateCourses.js:67]: image

For some reasons, I couldn't see the `UpdateCourses` button even though I was logged in as a Head Advisor. The DOM looks good to me, but I still couldn't see the button.

The console didn't throw any error.

**Review by philectron — CHANGES_REQUESTED** [review]
There are some minor changes needed ~~, also some questions that need to be answered~~ . Note that I skimmed over the client code; instead of inspecting it, I did some manual testing.

File: api/course.js (hunk 5 above)
Comment (silverware13) [reply api/course.js]: Good catch, you are correct that only the head advisor should be able to perform this action.

File: client/src/components/navbar/Navbar.js (hunk 6 above)
Comment (philectron) [reply client/src/components/navbar/Navbar.js:106]: I'm dumb. I opened the up-to-date version of your branch remotely on WSL and ran the out-of-date version on my Windows system. It's working.

File: client/src/components/navbar/UpdateCourses.js (hunk 7 above)
Comment (philectron) [reply client/src/components/navbar/UpdateCourses.js:67]: I'm dumb. I opened the up-to-date version of your branch remotely on WSL and ran the out-of-date version on my Windows system. It's working.

File: api/course.js (hunk 1 above)
Comment (silverware13) [reply api/course.js:30]: So the reason for no await is this: 

The operation is really slow, so slow (possibly hours) that it doesn't make sense to wait for the operation to finish to send a response (most requests will time out before we can get a response back). Instead the expected behavior is that we send a response to the user letting them know that we got their request and we are starting an update.

To let the user know the status of update we instead use notifications to signal when the update is done, this circumvents the issue of a request timing out after waiting for hour(s).

File: models/comment.js (hunk 3 above)
Comment (silverware13) [reply models/comment.js:23]: So I want to get your opinion here. 

If the getLiveCourses function has no await. Should the notifications it spawns have awaits?
I think logically they should only have awaits if we will perform some action if the notifications fail to send.

This raises the question, should we perform an action if they fail to send?
Should we cancel the update if the notification at the beginning or end of the update fails?

File: models/review.js (hunk 4 above)
Comment (silverware13) [reply models/review.js:24]: So I want to get your opinion here. 

If the getLiveCourses function has no await. Should the notifications it spawns have awaits?
I think logically they should only have awaits if we will perform some action if the notifications fail to send.

This raises the question, should we perform an action if they fail to send?
Should we cancel the update if the notification at the beginning or end of the update fails?

**Review by philectron — APPROVED** [review]
LGTM

**Pull request merged**